### DISCUSSION TITLE
[Edge Case Bug] Allow use of a premade UV2 instead of auto-unwrapping

### DISCUSSION THREAD

**Discussion by Omgesus** [godotengine/godot-proposals#5559] [3D]
Around Alpha 14 for Godot 4, there were some changes made to the UV2 unwrapper and/or the Lightmap/AO systems. After a while, I noticed that part of my issue was likely due to how the automatic unwrap works. In general, it works very well, but there's cases where it messes up, well... horribly. The following is how the automatic unwrap did a pillar, for example:

![2022-10-07-134207_912x967_scrot](https://user-images.githubusercontent.com/22787327/194718570-7112d2fd-4e07-4a0d-a0e9-5870e0ae5f6d.png)

As you can see, it's not ideal: the base and top flares (at the bottom of the UV unwrap) get broken up unnecessarily, and the pillar itself gets broken up into 4 segments unnecessarily. I went through and learned quickly how to add extra UV maps in Blender, then exported it as a GLTF and reimported it into Godot, expecting to be able to set the new UV map I made using a cylinder unwrap, but unfortunately found no such thing. 

While I hardly know any C/C++, I don't think this would necessarily be difficult to implement, and I would almost describe it as a necessity to create quality lighting. As stated, the automatic unwrap works well under _most_ circumstances, but when it doesn't, the results are something like this:

![2022-10-08-090138_3840x2160_scrot](https://user-images.githubusercontent.com/22787327/194718995-8311e008-2504-430f-88f0-b6cbb83b211a.png)

As long as there's no way to customize the unwrap settings, or better yet, use a premade UV2, I don't see a way to fix an issue like this. My proposal would be to, most likely, include the setting in the import chain. The following is a rough mockup of how I envision it being integrated:

![2022-10-08-090550_3072x1728_scrot](https://user-images.githubusercontent.com/22787327/194719634-0aed26d1-bb27-4380-a630-c7ec8acb683b.png)

The user would double click the file in the File System browser to access the import settings, and then click on the mesh. The import window has a header for it as well as a tick box. Selecting 'Enabled' then brings up a drop-down from which the user can select the UV map to use in place of the auto unwrap for UV2. 

At the end of the day, it isn't the single-most vital change, but I don't think it'd be the most difficult change to make, and would give intermediate to advanced end-users a better way to set up their meshes when the auto unwrap fails.

**Comment by Calinou** [comment]
This is likely a bug, as it [was possible in 3.x](https://docs.godotengine.org/en/stable/tutorials/3d/baked_lightmaps.html#unwrap-from-your-3d-dcc). Using premade UV2 *should* already happen if you don't specify the **Static Lightmaps** GI mode (but use **Static** instead).

Also, remember that not all 3D mesh formats can store UV2. For example, OBJ cannot store UV2, even if you import it as a *scene* (rather than as a *mesh*, which is the default).

**Reply by Omgesus** [reply]
Yep, it's a bug, I'll make a report in a few minutes if there isn't one! Using 'Automatic' for Lightmap UV generation (the default for importing) seems to be using enabled by default and overwriting UV2. I'll file a bug report in a few minutes.  Was a huge pain, taking literally a couple months, but finally solved my lighting issues with that. Now I just need to... update like 15 meshes to have a UV2 and all that... Fun :p


![2022-10-09-102932_3840x2160_scrot](https://user-images.githubusercontent.com/22787327/194773532-2338d75f-6ede-4435-9755-1b8e38b78ade.png)

**Reply by Omgesus** [reply]
> Yep, it's a bug, I'll make a report in a few minutes if there isn't one! Using 'Automatic' for Lightmap UV generation (the default for importing) seems to be using enabled by default and overwriting UV2. I'll file a bug report in a few minutes. Was a huge pain, taking literally a couple months, but finally solved my lighting issues with that. Now I just need to... update like 15 meshes to have a UV2 and all that... Fun :p
> 
> ![2022-10-09-102932_3840x2160_scrot](https://user-images.githubusercontent.com/22787327/194773532-2338d75f-6ede-4435-9755-1b8e38b78ade.png)

Odd... I was making a minimal project for it, and it _did_ use the second UV automatically. Why am I the one that keeps running into bugs that happen only in my project... Probably just bugs due to transferring the project since pre-alpha days :p So I guess it _is_ a bug, but only happens under certain edge cases that I have no clue what the causes are. At least the issue is somewhat solved, then.
